csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -742,11 +747,18 @@ private module Cached {
     getANonTypeParameterSubType(t1) = getANonTypeParameterSubType(t2)
   }
 
+  /**
+   * Holds if GVNs `t1` and `t2` may have a common sub type. Neither `t1` nor
+   * `t2` are allowed to be type parameters.
+   */
+  cached
+  predicate commonSubType(DataFlowType t1, DataFlowType t2) { commonSubTypeGeneral(t1, t2) }
+
   cached
   predicate commonSubTypeUnifiableLeft(DataFlowType t1, DataFlowType t2) {
-    exists(DataFlowType t |
+    exists(Gvn::GvnType t |
       Gvn::unifiable(t1, t) and
-      commonSubType(t, t2)
+      commonSubTypeGeneral(t, t2)
```

Comment:
As a general comment here: a lot of these annotations seem to be working around actual/supposed compiler missteps. For example, here one might more naturally use `commonSubType`, but `pragma[nomagic]` suggests that there has been a compilation problem that is being worked around. This leads to the code being more verbose and split up in slightly unnatural ways. Is there a better way to document what's being worked around, or should I simply continue to "filter out" all of these pragmas in my mind?

Reply:
Actually, here we can't use `commonSubType`, because we need the more general variant where the first column is not restricted to `DataFlowType` (but the more general `DataFlowTypeOrUnifiable` type).